doc/string/slices.rdoc

```diff
@@ -0,0 +1,144 @@
+== \String Slices
+
+A _slice_ of a string is a substring that is selected by certain criteria.
+
+These instance methods make use of slicing:
+
+- String#[] (and its alias String#slice) return a slice copied from +self+.
+- String#[]= returns a copy of +self+ with a slice replaced.
+- String#slice! returns +self+ with a slice replaced.
+
+Each of the above methods takes arguments that determine the slice
+to be copied or replaced.
+
+The arguments have several forms.
+For string +string+,  the forms are:
+
+- <tt>string[index]</tt>.
+- <tt>string[start, length]</tt>.
+- <tt>string[range]</tt>.
+- <tt>string[regexp, capture = 0]</tt>.
+- <tt>string[substring]</tt>.
+
+<b><tt>string[index]</tt></b>
+
+When non-negative integer argument +index+ is given,
+the slice is the 1-character substring found in +self+ at character offset +index+:
+
+  'bar'[0]       # => "b"
+  'bar'[2]       # => "r"
+  'bar'[20]      # => nil
+  'тест'[2]      # => "с"
+  'こんにちは'[4]  # => "は"
+
+When negative integer +index+ is given,
+the slice begins at the offset given by counting backward from the end of +self+:
+
+  'bar'[-3]         # => "b"
+  'bar'[-1]         # => "r"
+  'bar'[-20]        # => nil
+
+<b><tt>string[start, length]</tt></b>
+
+When non-negative integer arguments +start+ and +length+ are given,
+the slice begins at character offset +start+, if it exists,
+and continues for +length+ characters, if available:
+
+  'foo'[0, 2]       # => "fo"
+  'тест'[1, 2]      # => "ес"
+  'こんにちは'[2, 2]  # => "にち"
+  # Zero length.
+  'foo'[2, 0]       # => ""
+  # Length not entirely available.
+  'foo'[1, 200]     # => "oo"
+  # Start out of range.
+  'foo'[4, 2]      # => nil
+  'foo'[-4, 2]     # => nil
+
+Special case: if +start+ is equal to the length of +self+,
+the slice is a new empty string:
+
+  'foo'[3, 2]   # => ""
+  'foo'[3, 200] # => ""
+
+When negative +start+ and non-negative +length+ are given,
+the slice beginning is determined by counting backward from the end of +self+,
+and the slice continues for +length+ characters, if available:
+
+  'foo'[-2, 2]    # => "oo"
+  'foo'[-2, 200]  # => "oo"
+
+When negative +length+ is given, there is no slice:
+
+  'foo'[1, -1]  # => nil
+  'foo'[-2, -1] # => nil
+
+<b><tt>string[range]</tt></b>
+
+When Range argument +range+ is given,
+derives +start+ and +length+ values from the given +range+:
+
+  start = range.begin
+  length = range.end - range.begin + 1 # May be different from range.size.
+  length = length < 0 ? 0 : length     # Length is never negative.
+
+The slice is then determined as above:
```

Comment:
I think this is hard to understand and not correct when the range doesn't include the end.

For example:

```ruby
irb(main):033:0> r = 0..1
=> 0..1
irb(main):034:0> r.end
=> 1
irb(main):035:0> "hello"[r]
=> "he"
irb(main):036:0> r = 0...1
=> 0...1
irb(main):037:0> r.end
=> 1
irb(main):038:0> "hello"[r]
=> "h"
```

I suggest something like the following:

```suggestion
creates a substring of +string+ using the indices in +range+.
```

Reply:
Done.